pkg/tsdb/cloudwatch/time_series_query.go

```diff
@@ -5,120 +5,105 @@ import (
-
 	resp := backend.NewQueryDataResponse()
 
-	for _, q := range req.Queries {
-		startTime := q.TimeRange.From
-		endTime := q.TimeRange.To
-		if !startTime.Before(endTime) {
-			return nil, fmt.Errorf("invalid time range: start time must be before end time")
-		}
-
-		requestQueriesByRegion, err := e.parseQueries(req, startTime, endTime)
-		if err != nil {
-			return nil, err
-		}
-
-		if len(requestQueriesByRegion) == 0 {
-			return backend.NewQueryDataResponse(), nil
-		}
-
-		resultChan := make(chan *backend.DataResponse, len(req.Queries))
-		eg, ectx := errgroup.WithContext(ctx)
-		for r, q := range requestQueriesByRegion {
-			requestQueries := q
-			region := r
-			eg.Go(func() error {
-				defer func() {
-					if err := recover(); err != nil {
-						plog.Error("Execute Get Metric Data Query Panic", "error", err, "stack", log.Stack(1))
-						if theErr, ok := err.(error); ok {
-							resultChan <- &backend.DataResponse{
-								Error: theErr,
-							}
-						}
-					}
-				}()
+	if len(req.Queries) == 0 {
+		return nil, fmt.Errorf("request contains no queries")
+	}
 
-				client, err := e.getCWClient(region, req.PluginContext)
-				if err != nil {
-					return err
-				}
+	// startTime and endTime are always the same for all queries
+	startTime := req.Queries[0].TimeRange.From
+	endTime := req.Queries[0].TimeRange.To
+	if !startTime.Before(endTime) {
+		return nil, fmt.Errorf("invalid time range: start time must be before end time")
+	}
 
-				queries, err := e.transformRequestQueriesToCloudWatchQueries(requestQueries)
-				if err != nil {
-					for _, query := range requestQueries {
-						resultChan <- &backend.DataResponse{
-							Frames: data.Frames{data.NewFrame(query.RefId)},
-							Error:  err,
-						}
-					}
-					return nil
-				}
+	requestQueriesByRegion, err := e.parseQueries(req.Queries, startTime, endTime)
+	if err != nil {
+		return nil, err
+	}
 
-				metricDataInput, err := e.buildMetricDataInput(startTime, endTime, queries)
-				if err != nil {
-					return err
-				}
+	if len(requestQueriesByRegion) == 0 {
+		return backend.NewQueryDataResponse(), nil
+	}
 
-				cloudwatchResponses := make([]*cloudwatchResponse, 0)
-				mdo, err := e.executeRequest(ectx, client, metricDataInput)
-				if err != nil {
-					for _, query := range requestQueries {
-						resultChan <- &backend.DataResponse{
-							Frames: data.Frames{data.NewFrame(query.RefId)},
-							Error:  err,
+	resultChan := make(chan *responseWrapper, len(req.Queries))
+	eg, ectx := errgroup.WithContext(ctx)
+	for r, q := range requestQueriesByRegion {
+		requestQueries := q
+		region := r
+		eg.Go(func() error {
+			defer func() {
+				if err := recover(); err != nil {
+					plog.Error("Execute Get Metric Data Query Panic", "error", err, "stack", log.Stack(1))
+					if theErr, ok := err.(error); ok {
+						resultChan <- &responseWrapper{
+							DataResponse: &backend.DataResponse{
+								Error: theErr,
+							},
```

Comment:
is it okay to have `refID: 0` here?

Reply:
Yeah I think it's okay. Or, I mean the real error that caused the panic, is not passed all the way to the panel since the error is not attached to a certain ref id. The panel just receives a general error. But I think that's fine.